### PR TITLE
Add some Simple-PicoLM features

### PR DESCRIPTION
**Pull request by HosLak**
This PR introduces several improvements inspired by Simple-PicoLM (my previous Blueberry repo):

- Added DeepSeek-warmup strategy
- GQA
- Smart & auto d_ff initialization
- Applied QK norm
- Set dropout = 0
- Initialized w_o = 0 in llm.py
- Set n_layer = 8, n_experts = 6

Results comparison:

  Before:
  - Total parameters: 79,059,840
  - Step 500 → Val Loss: 3.5751, Val Acc: 0.3193, Val PPL: 35.70
  - Step 1000 → Val Loss: 0.2550, Val Acc: 0.9537, Val PPL: 1.29
  
  After:
  - Total parameters: 74,178,048
  - Step 500 → Val Loss: 2.7484, Val Acc: 0.4863, Val PPL: 15.62
  - Step 1000 → Val Loss: 0.1293, Val Acc: 0.9816, Val PPL: 1.14

Note: Training speed is slightly slower with `n_layer=8`.  
If speed matters more, `n_layer=6` works fine too, but I recommend 8 layers since the stability and performance gains are worth it.

### CODE REVIEW

File: legacy/llm.py [hunk 1]
```diff
@@ -142,7 +150,7 @@ def load_and_cache_data(config: MoEModelConfig, cache_dir: str = "data_cache"):
     print(f"🔄 Processing new data (will cache for future use)")
 
     # Load tokenizer
-    tokenizer = AutoTokenizer.from_pretrained("HuggingFaceTB/SmolLM-135M", token=False)
+    tokenizer = AutoTokenizer.from_pretrained("openai-community/gpt2", token=False)
```
Comment (vukrosic) [inline legacy/llm.py:153]: wny reason to use gpt2 tokenizer over smollm? i think you are still using smollm dataset since i don't see any changes here

**Review by vukrosic — COMMENTED** [review]
is this the performance comparison:

Before:
📊 Total parameters: 79,059,840
Step 500 → Val Loss: 3.5751, Val Acc: 0.3193, Val PPL: 35.70
Step 1000 → Val Loss: 0.2550, Val Acc: 0.9537, Val PPL: 1.29

After:
📊 Total parameters: 74,178,048
Step 500: Val Loss: 2.7420, Val Acc: 0.4905, Val PPL: 15.52
Step 1000: Val Loss: 0.1287, Val Acc: 0.9816, Val PPL: 1.14

File: legacy/llm.py (hunk 1 above)
Comment (HosLak) [reply legacy/llm.py:153]: Ah, my bad, this was a mistake on my part  
I originally thought SmolLM tokenizer had a very large vocab (like 128k), but I just realized I was wrong. In fact, SmolLM’s vocab is actually a bit smaller than GPT-2's, and it’s nicely aligned to power-of-two sizes (like 128), which makes training even faster.
So yeah, it's definitely better to stick with SmolLM tokenizer.  
Sorry about the confusion

**Comment by HosLak** [comment]
Recent Updates Pushed:

- I enabled bias in Q, K, V projections (bias=True), so now this part is aligned with modern best practices.
- I also applied torch.compile() only during training, to keep full compatibility.

Why I used torch.compile() only for training:
Some GPUs don’t support torch.compile(), but the user may still want to run inference.
So what I did is: compile only during training, then save the model without compilation.
This way the checkpoint works everywhere for inference with no issues.

Performance Impact:
After these changes, training speed went from about ~1.8 it/s → ~2.7 it/s.
That’s roughly a 50% speedup.
There was zero negative impact on model performance, actually, thanks to qkv bias=True, the model is performing even better now.

Training Time:
before: 19:20
after: 13:56

You might think with speed going from 1.8 → 2.7 it/s, the training time should drop to ~10 minutes.
But in reality it’s ~14 minutes.
The reason is: before training actually starts, the model “hangs” for a little bit (not sure why yet).
Hopefully I’ll figure out the cause soon, because fixing it would save those few extra minutes.

**Comment by vukrosic** [comment]
@HosLak can you just change pull request to merge into hossein-llm branch instead of main so i can test it all there first

also i think you don't need ot make changes to legacy/llm.py anymore, i might remove it, unless you think otherwise?

**Comment by HosLak** [comment]
> also i think you don't need ot make changes to legacy/llm.py anymore, i might remove it, unless you think otherwise?

For now, don’t rush to remove llm.py,
’cause a couple of my changes are still only applied there.

**Comment by HosLak** [comment]
I’ve closed this pull request for now.
But I’ll be opening a few new ones soon - this time I won’t include everything in a single PR.
I’ll make the changes one by one, so you can decide which ones to merge and which ones to skip.
(Also, this time I won’t touch llm.py, so you can safely remove it if you want.)
